### DISCUSSION TITLE
Runtime Types in pyo3

### DISCUSSION THREAD

**Discussion by vkobinski** [PyO3/pyo3#4271] [Questions]
I want to implement a match case for a Type that is parsed at Runtime, i just need to match to the right case when the match is executed, I achieved a state where I can desestructure the arguments, but I can't make the right case be matched.

My code:
https://github.com/vkobinski/benda-main/blob/1405203f697ba3232160943cf4b36719134e432b/crates/benda/src/types/book.rs#L164

My problem:
All the types i create at runtime are of the type builtins.Ctr, i just need to change this type for the match case.

**Comment by davidhewitt** [comment]
Can you reduced this down to a more contained example of the problem? I took a quick look at your link but there's a lot there and it's not immediately clear to me what you're asking.

**Reply by vkobinski** [reply]
For a match statement to work, I must return a PyType for every case. But, the only PyClass that I define is Ctr, so any object I try to match ends up being matched to the first case because they are all Ctr.
There is some way to create a PyType at runtime, or change its name so that the match case works?

**Reply by vkobinski** [reply]
I'm trying to map types from another programming language into Python, but these types are all mapped at Runtime, there is a way to implement a PyClass at runtime?

**Reply by davidhewitt** [reply]
`PyTypeInfo` as a trait is a compile-time concept.

You're probably going to need to work with `Bound<PyType>` and make runtime decisions.

**Reply by davidhewitt** [reply]
Can you write out in short form the `match` you're trying to implement? Maybe I can offer suggestions then.

**Reply by vkobinski** [reply]
Do you know how Python compares an object with a Type? I tried changing the PyTypeObject name and it continued to be matched.

**Reply by vkobinski** [reply]
```
my_list = book.adts.List.Cons(10, book.adts.List.Nil())

while my_list:
    match my_list:
        case book.adts.List.tCons(value, tail):
            print(value)
            lista = tail
        case book.adts.List.tNil():
            print("End")
            break
```

And the type is returned by this PyClass:
```
#[pyclass(name = "Ctrs")]
#[derive(Clone, Debug, Default)]
pub struct Ctrs {
    fields: IndexMap<String, Ctr>,
}

#[pymethods]
impl Ctrs {
    // TODO: Return Type or Constructor

    fn __getattr__(&self, object: Bound<PyAny>) -> PyResult<pyo3::PyObject> {
        let py = object.py();

        if object.to_string().starts_with("t") {
            let b_name = object.to_string();
            let name = b_name.strip_prefix("t").unwrap();

            let type_obj = Ctr::type_object_bound(py);

            return Ok(type_obj.to_object(py));
        }

        if let Some(val) = self.fields.get(&object.to_string()) {
            Ok(Py::new(py, val.clone()).unwrap().into_any())
        } else {
            new_err(format!("Could not find attr {}", object))
        }
    }
}
```

But all objects like this are Ctr, i just need to make List.Cons not map to Ctr.

**Reply by davidhewitt** [reply]
I think probably you want to use [`object.is_instance(cons)`](https://docs.rs/pyo3/latest/pyo3/prelude/trait.PyAnyMethods.html#tymethod.is_instance)? - Where you can use `py.import` to load the Python module and get the `Cons` type.

**Reply by vkobinski** [reply]
I'm trying to modify isinstace for my Ctr PyClass, but I'm not able to find what method Python is calling on my object. Pretty hard.

**Comment by vkobinski** [comment]
```
while my_list:
    match my_list.name:
        case book.adts.List.Cons.name:
            print(my_list.head)
            my_list = my_list.tail
        case book.adts.List.Nil.name:
            print("End")
            break
```

My solution for now, can't make the PyType work.
